### PR TITLE
Distributed optimizer support for BERT

### PR DESCRIPTION
**Pull request by timmoon10**
# What does this PR do ?

Add support for the distributed Adam optimizer to BERT.

**Collection**: NLP

# Changelog 
- Add support for the Apex distributed Adam optimizer to BERT

# Usage

Set the optimizer to `distributed_fused_adam` in the config file:

https://github.com/NVIDIA/NeMo/blob/79a22eaee0813d5a0e4dc82ebd4a43071399c899/examples/nlp/language_modeling/conf/megatron_bert_config.yaml#L118

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

Pinging @shanmugamr1992.

# Additional Information
* Depends on https://github.com/NVIDIA/NeMo/pull/5293.
* BERT distopt support in this PR is basically identical to GPT and T5 distopt support in https://github.com/NVIDIA/NeMo/pull/5236.

### CODE REVIEW

**Comment by timmoon10** [comment]
Running on 4 A100s for 20 steps (in BF16 with O2 optimizations):

   | ZeRO | DP | PP | TP | Throughput | Train loss | Val loss | Memory usage |
   |------|----|----|----|------------|------------|----------|-----------|
   | No   |  1 |  1 |  1 | 4.50it/s   |       10.2 |    9.850   | 4002 MB     |
   | Yes  |  1 |  1 |  1 | 4.61it/s   |       10.2 |    9.850 | 4052 MB     |
   | No   |  2 |  1 |  1 | 3.69it/s   |       10.2 |    9.800   | 4002 MB     |
   | Yes  |  2 |  1 |  1 | 3.81it/s   |       10.2 |    9.790   | 3432 MB     |
   | No   |  4 |  1 |  1 | 3.09it/s   |       10.1 |    9.700   | 4002 MB     |
   | Yes  |  4 |  1 |  1 | 3.16it/s   |       10.1 |    9.690    | 3134 MB     |
   | No   |  2 |  2 |  1 | 1.99it/s   |       10.1 |    9.660   | 2580 MB     |
   | Yes  |  2 |  2 |  1 | 1.94it/s   |       10.1 |    9.660   | 2540 MB     |
   | No   |  2 |  1 |  2 | 2.26it/s   |       10.1 |    9.740   | 2358 MB     |
   | Yes  |  2 |  1 |  2 | 2.22it/s   |       10.1 |    9.730   | 2174 MB     |                                           

I think this PR is good to go.

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
